extend converter interface and add additional time related converters

slu-it opened the pull request:
**A lot of changes and additions to the type converters:**

(1) The `Field` or `Parameter` (`AnnotatedElement`) for which a value is converted, is now provided to the converter in order to allow for annotation-based configuration of the conversion. E.g. by using `@Pattern` to provide the pattern for date interpretation.

(2) Additional converters for LocalTime, LocalDateTime and OffsetDateTime.

(3) Format pattern for parsing time related classes can now be overridden by annotating the target with `@Pattern`.

(4) Moved time related converters to own sub-package `time`.

slu-it commented:
I suggest, after merging this, the next step would be to move all the implementation classes to a separate module. The API module should only contain annotations, interfaces and, exceptions. The type converter implementations belong either in the engine module or in a separate module altogether.

bitterblue commented on livingdoc-fixture-api/src/main/java/org/livingdoc/fixture/api/converter/predicate/AbstractNumberPredicateConverter.java:
```diff
@@ -0,0 +1,55 @@
+
+    private String toScript(T number, String expression) {
+        String script = expression;
+        script = script.replaceAll("([^\\|])\\|([^\\|])", "$1||$2");
```
This won't work if I use a more complex expression, e.g. `(x > 1) && (x < 2) || (x > 3) && (x  < 4)`.

bitterblue commented on livingdoc-fixture-api/src/main/java/org/livingdoc/fixture/api/converter/predicate/AbstractNumberPredicateConverter.java:
```diff
@@ -0,0 +1,55 @@
+        String script = expression;
+        script = script.replaceAll("([^\\|])\\|([^\\|])", "$1||$2");
+        script = script.replaceAll("([^\\&])\\&([^\\&])", "$1&&$2");
+        script = script.replaceAll("x", String.valueOf(number));
```
Variable substitution by regex. This is so going to break :D

slu-it commented:
Removed the Predicate commit from this PR. I'll push it to my sandbox on a separate branch. Then we can discuss further.

slu-it commented:
@bitterblue updated PR

I refactored the time related converters to use a common base class (90% of their code) and split the known exceptions into two: 

1. `ValueFormatException` (specialized `ConversionException`) indicating that the value does not match the intended format.
2. `MalformedFormatException` indicating, that the custom format is malformed.

bitterblue commented:
Since you removed the duplication from the converters, why not refactor the unit tests as well?

slu-it commented:
There is only one test method which is kind of redundant: `defaultFormatAssumedIfNoElementGiven`
All the other test methods depend on actual types.

Additionally I like to make my unit tests as implementation agnostic as possible. We could for sure test the principle of defining a custom format within a test class for the  `AbstractTemporalConverter`, but that would couple the tests unnecessarily to the implementation details of the converters (being sub-classes of `AbstractTemporalConverter`). This way the tests will be stable regardless of any refactorings done to the classes under test.

When created the `AbstractTemporalConverter` class, I didn't have to change a single test. If we ever need to change it back, or split the implementation further, I would like it to be the same.

slu-it commented:
@pkleimann or @Drakojin could one of you merge this (3/4 votes required :))
